### PR TITLE
Add TE MXFP8 multi-stream implementation in TorchAO

### PR DESCRIPTION
**Pull request by elfiegg**
To-dos:
1. Microbenchmark the kernel performance compared to TorchAO cutlass
2. More holistic verification on workloads like DeepseekV3 and FLUX

Previously observed 1.24x speedup to e2e DeepseekV3 workload, see description: https://github.com/pytorch/torchtitan/pull/2207

### CODE REVIEW

File: torchao/prototype/moe_training/config.py [hunk 1]
```diff
@@ -129,6 +130,13 @@ def from_recipe(
                 wgrad_with_hp=False,
                 scale_calculation_mode=ScaleCalculationMode.RCEIL,
             )
+        elif recipe == MXFP8TrainingRecipe.MXFP8_TE:
+            return cls(
```
Comment (vkuzo) [inline torchao/prototype/moe_training/config.py]: how about adding a `__post_init__` method and asserting that TransformerEngine is installed there, and linking to the docs of how to install it? This way the user can go install TE as early as possible, before they even try to feed data through the model.

File: test/prototype/moe_training/test_te_grouped_mm.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+from torchao.utils import torch_version_at_least
+
+if not (
+    torch_version_at_least("2.7.0")
+    and torch.cuda.is_available()
+    and torch.cuda.get_device_capability()[0] >= 9
+):
+    pytest.skip("Unsupported PyTorch version or hardware", allow_module_level=True)
+
+try:
```
Comment (vkuzo) [inline test/prototype/moe_training/test_te_grouped_mm.py:31]: I think we should try running this in CI, as long as reliability is good.  Doesn't have to be in this PR and torchao team can help with that.

File: test/prototype/moe_training/test_te_grouped_mm.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        f"Weight grad sqnr {sqnr_weight_grad} too low"
+    )
+
+
+# ──────────────────────────────────────────────────────────────────────────────
+# Test config / recipe
+# ──────────────────────────────────────────────────────────────────────────────
+
+
+def test_mxfp8_te_recipe():
```
Comment (vkuzo) [inline test/prototype/moe_training/test_te_grouped_mm.py]: nit: delete? i dont think this needs a test

File: test/prototype/moe_training/test_te_grouped_mm.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+# ──────────────────────────────────────────────────────────────────────────────
+# Test torch.compile compatibility (fake impls)
+# ──────────────────────────────────────────────────────────────────────────────
+
+
+@pytest.mark.parametrize("M,K,N", [(1024, 1024, 1024)])
+@pytest.mark.parametrize("num_experts", (4,))
+def test_te_compile(M, K, N, num_experts):
+    """Verify that torch.compile traces through TE custom ops without graph breaks."""
```
Comment (vkuzo) [inline test/prototype/moe_training/test_te_grouped_mm.py:248]: nice!

File: torchao/prototype/moe_training/te_grouped_mm.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+    real rows and zero-fills padding rows in a single kernel launch.
+    """
+    padded_splits = [_ceil_to_block(m) for m in m_splits]
+    if padded_splits == m_splits:
+        return tensor.contiguous(), m_splits
+
+    padded_total = sum(padded_splits)
+    K = tensor.shape[-1]
+    padded = torch.empty(padded_total, K, dtype=tensor.dtype, device=tensor.device)
+    tex.fused_multi_row_padding(
```
Comment (danielvegamyhre) [inline torchao/prototype/moe_training/te_grouped_mm.py:99]: quick questions:

1. does this padding kernel have logic to avoid the extra copy overhead if the tokens are already aligned (e.g., if HybridEP already aligned tokens during dispatch)?
2. do you have any microbenchmarks for this fused padding op? i am working on a triton kernel for this and found it trickier than expected to get good performance with triton for this, so am curious about the memory bandwidth utilization % of this cuda (?) kernel

File: torchao/prototype/moe_training/DESIGN.md [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+       ├─ forward:  te_moe::gemm_fwd   (TE MXFP8Quantizer → general_grouped_gemm)
+       ├─ dgrad:    te_moe::gemm_dgrad  (TE MXFP8Quantizer → general_grouped_gemm)
+       └─ wgrad:    te_moe::gemm_wgrad  (TE MXFP8Quantizer → general_grouped_gemm)
+```
+
+---
+
+## 4. File Changes
+
+### 4.1 `torchao/quantization/quantize_/common/kernel_preference.py`
```
Comment (danielvegamyhre) [inline torchao/prototype/moe_training/DESIGN.md]: this md file reads like a plan for agents but i actually think it might be useful to land in some form, without the list of file changes here. i have been wanting to improve documentation for this feature in torchao and this would help.

File: test/prototype/moe_training/test_te_grouped_mm.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+  5. torch.compile compatibility (custom ops have fake impls)
+"""
+
+import pytest
+import torch
+from torch.nn import functional as F
+
+from torchao.utils import torch_version_at_least
+
+if not (
```
Comment (danielvegamyhre) [inline test/prototype/moe_training/test_te_grouped_mm.py:24]: Can you also update the benchmark script here to include this implementation? https://github.com/pytorch/ao/blob/d6d423e0322b760cf8c5d69934915a837f319a99/benchmarks/prototype/moe_training/benchmark_scaled_grouped_mm_dq.py#L66

File: torchao/prototype/moe_training/config.py [hunk 8]
```diff
@@ -31,6 +31,7 @@ class MXFP8TrainingRecipe(Enum):
     MXFP8_RCEIL = "mxfp8_rceil"
     MXFP8_RCEIL_WGRAD_WITH_HP = "mxfp8_rceil_wgrad_with_hp"
     MXFP8_EMULATED_RCEIL = "mxfp8_emulated_rceil"
+    MXFP8_TE = "mxfp8_te"
```
Comment (vkuzo) [inline torchao/prototype/moe_training/config.py]: Currently `MXFP8TrainingRecipe` is purely to specify numerics and not the kernels.  The current PR introduces a kernel level optimization, IMO cleaner to have that be a separate knob.

Comment (vkuzo) [reply torchao/prototype/moe_training/config.py]: I think if you name it something like `MXFP8_RCEIL_WITH_TE_KERNELS` it would be cleaner

File: test/prototype/moe_training/test_te_grouped_mm.py (hunk 2 above)
Comment (elfiegg) [reply test/prototype/moe_training/test_te_grouped_mm.py:31]: Sounds good. To add this in CI, TransformerEngine and its dependencies need to be in the container. The CI docker image will need to install that. I can help with it, but of course only if needed :)

File: test/prototype/moe_training/test_te_grouped_mm.py (hunk 3 above)
Comment (elfiegg) [reply test/prototype/moe_training/test_te_grouped_mm.py]: Done

File: test/prototype/moe_training/test_te_grouped_mm.py (hunk 4 above)
Comment (elfiegg) [reply test/prototype/moe_training/test_te_grouped_mm.py:248]: c:

File: torchao/prototype/moe_training/config.py (hunk 1 above)
Comment (elfiegg) [reply torchao/prototype/moe_training/config.py]: Good idea

File: test/prototype/moe_training/test_te_grouped_mm.py (hunk 7 above)
Comment (elfiegg) [reply test/prototype/moe_training/test_te_grouped_mm.py:24]: Done

File: torchao/prototype/moe_training/te_grouped_mm.py (hunk 5 above)
Comment (elfiegg) [reply torchao/prototype/moe_training/te_grouped_mm.py:99]: 1.  Yes, line 91 - 94 will exit if no padding is needed for all experts. But if per-split (per-expert) is not aligned, TE's kernel will still need to include in the kernel launch and copy it to the output buffer: https://github.com/NVIDIA/TransformerEngine/blob/d40b9de312c3cacec73b10a71b8d8419a11824c8/transformer_engine/pytorch/csrc/extensions/padding.cpp#L61-L82 
2. Added one benchmark for that. Please check it out in file `bench_te_fused_padding.py`
3. From the benchmark, TE fused padding is ~2.2-2.6x faster than naive (torch.cat + F.pad) for unaligned/random splits across all sizes. Unpad is the opposite story: TE unpad is consistently slower (~0.73-0.83x) than naive even for random splits. See some raw data below
It's not super impressive, a kernel using vectorized loads (128-bit float4 or wider) and removing the per-element bounds checks (for arbitrary K, which we don't have) could potentially push to 80-85%
## Padding (TE `fused_multi_row_padding` vs naive `F.pad` + `cat`)

| M | K | Experts | TE (us) | Naive (us) | Speedup | BW (GB/s) | BW% |
|------:|-----:|--------:|--------:|-----------:|--------:|----------:|----:|
| 16384 | 1536 | 1 | 29.73 | 75.10 | 2.53x | 2916.5 | 36.5% |
| 16384 | 1536 | 4 | 87.09 | 191.33 | 2.20x | 4227.2 | 52.8% |
| 16384 | 1536 | 8 | 150.88 | 371.71 | 2.46x | 4811.4 | 60.1% |
| 16384 | 1536 | 16 | 287.74 | 745.79 | 2.59x | 5061.5 | 63.3% |
| 16384 | 2048 | 1 | 33.79 | 59.68 | 1.77x | 3421.0 | 42.8% |
| 16384 | 2048 | 4 | 111.62 | 236.54 | 2.12x | 4397.7 | 55.0% |
| 16384 | 2048 | 8 | 199.68 | 459.07 | 2.30x | 4847.3 | 60.6% |
| 16384 | 2048 | 16 | 381.95 | 915.76 | 2.40x | 5084.1 | 63.6% |
| 16384 | 5120 | 1 | 68.67 | 140.26 | 2.04x | 4208.5 | 52.6% |
| 16384 | 5120 | 4 | 242.98 | 562.32 | 2.31x | 5050.5 | 63.1% |
| 16384 | 5120 | 8 | 464.22 | 1099.78 | 2.37x | 5212.6 | 65.2% |
| 16384 | 5120 | 16 | 914.64 | 2199.52 | 2.40x | 5307.8 | 66.3% |
| 16384 | 7168 | 1 | 91.23 | 190.34 | 2.09x | 4434.9 | 55.4% |
| 16384 | 7168 | 4 | 340.99 | 770.50 | 2.26x | 5038.2 | 63.0% |
| 16384 | 7168 | 8 | 656.74 | 1506.32 | 2.29x | 5158.4 | 64.5% |
| 16384 | 7168 | 16 | 1299.49 | 3019.15 | 2.32x | 5230.2 | 65.4% |

## Unpadding (TE `fused_multi_row_unpadding` vs naive `split` + `cat`)

| M | K | Experts | TE (us) | Naive (us) | Speedup | BW (GB/s) | BW% |
|------:|-----:|--------:|--------:|-----------:|--------:|----------:|----:|
| 16384 | 1536 | 1 | 29.66 | 29.18 | 0.98x | 2922.8 | 36.5% |
| 16384 | 1536 | 4 | 87.07 | 59.71 | 0.69x | 4228.0 | 52.9% |
| 16384 | 1536 | 8 | 150.78 | 110.59 | 0.73x | 4814.4 | 60.2% |
| 16384 | 1536 | 16 | 288.13 | 210.94 | 0.73x | 5054.8 | 63.2% |
| 16384 | 2048 | 1 | 33.79 | 25.60 | 0.76x | 3421.0 | 42.8% |
| 16384 | 2048 | 4 | 111.58 | 77.86 | 0.70x | 4399.0 | 55.0% |
| 16384 | 2048 | 8 | 199.68 | 143.68 | 0.72x | 4847.3 | 60.6% |
| 16384 | 2048 | 16 | 381.95 | 279.26 | 0.73x | 5084.1 | 63.6% |
| 16384 | 5120 | 1 | 68.67 | 54.85 | 0.80x | 4208.5 | 52.6% |
| 16384 | 5120 | 4 | 244.70 | 184.61 | 0.75x | 5014.8 | 62.7% |
| 16384 | 5120 | 8 | 465.70 | 352.26 | 0.76x | 5196.1 | 65.0% |
| 16384 | 5120 | 16 | 915.52 | 695.41 | 0.76x | 5302.7 | 66.3% |
| 16384 | 7168 | 1 | 91.14 | 71.39 | 0.78x | 4439.6 | 55.5% |

File: torchao/prototype/moe_training/config.py (hunk 8 above)
Comment (elfiegg) [reply torchao/prototype/moe_training/config.py]: Removed it entirely from recipe for now and downstream needs to specifically use KernelPreference=TE to initialize a MXFP8TrainingOpConfig.

File: torchao/prototype/moe_training/te_grouped_mm.py (hunk 5 above)
Comment (danielvegamyhre) [reply torchao/prototype/moe_training/te_grouped_mm.py:99]: this is great, thanks!

File: torchao/prototype/moe_training/DESIGN.md [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+  → _MXFP8GroupedMM.forward(kernel_preference=TE)
+  → te_moe::gemm_fwd
+```
+
+---
+
+## 6. Known Caveats
+
+1. **GPU→CPU sync**: `_offs_to_m_splits` calls `offs.tolist()` which is a
+   device-to-host synchronization. This happens 3 times per MoE layer per
```
Comment (danielvegamyhre) [inline torchao/prototype/moe_training/DESIGN.md:280]: i am surprised to read this - can you share the reasoning behind the decision to require a d2h sync for each of the forward, dgrad, and wgrad? this seems like it would kill overall training throughput as it keeps halting and waiting for the sync before resuming kernel dispatches.

why doesn't the grouped_mm accept the m_splits as a tensor on the device? this is the way `torch._scaled_grouped_mm` works so i am curious. 

The perf numbers you shared look good so i am curious about the details of this
